setup-env

```diff
@@ -4,53 +4,199 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
     exit 1
 fi
 
+PROGNAME="setup-env"
+
 if [ -n "$BASH_SOURCE" ]; then
-    OEROOT="`dirname $BASH_SOURCE`"
+    CWD="`dirname $BASH_SOURCE`"
 elif [ -n "$ZSH_SOURCE"]; then
-    OEROOT="`dirname $0`"
+    CWD="`dirname $0`"
 else
-    OEROOT="`pwd`"
+    CWD="`pwd`"
 fi
+
 if [ -n "$BBSERVER" ]; then
     unset BBSERVER
 fi
-if [ "x$1" = "x" ]; then
-    BDIR="build"
-else
-    BDIR="$1"
-fi
-if [ -e "$BDIR" ]; then
-    BDIR=`readlink -f "$BDIR"`
-    if [ ! -d "$BDIR" ]; then
-	echo "Usage: setup-env [dirname]"
-	return 1
+
+usage()
+{
+    cat <<EOF
+Usage: MACHINE=<machine> DISTRO=<distro> source $PROGNAME <build-dir>
+Usage:                                   source $PROGNAME <build-dir>
+    <machine>    machine name
+    <distro>     distro name
+    <build-dir>  build directory
+The first usage is for creating a new build directory. In this case, the
+script creates the build directory <build-dir>, configures it for the
+specified <machine> and <distro>, and prepares the calling shell for running
+bitbake on the build directory.
+The second usage is for using an existing build directory. In this case,
+the script prepares the calling shell for running bitbake on the build
+directory <build-dir>. The build directory configuration is unchanged.
+
+EOF
+    ls layers/*/conf/machine/*.conf > /dev/null 2>&1
+    ls layers/meta-testdistro/conf/distro/*.conf > /dev/null 2>&1
+    if [ $? -eq 0 ]; then
+        cat <<EOF
+Supported machines: `echo; ls layers/*/conf/machine/*.conf \
+| sed s/\.conf//g | sed -r 's/^.+\///' | xargs -I% echo -e "\t%"`
+Supported Tegra test distros: `echo; ls layers/meta-testdistro/conf/distro/*.conf \
+| sed s/\.conf//g | sed -r 's/^.+\///' | xargs -I% echo -e "\t%"`
+Examples:
+- To create a new Yocto build directory:
+  $ MACHINE=jetson-tx2 DISTRO=testdistro source $PROGNAME build
+- To use an existing Yocto build directory:
+  $ source $PROGNAME build
+
+EOF
     fi
+}
+
+clean_up()
+{
+   unset LIST_MACHINES VALID_MACHINE
+   unset CWD TEMPLATES SHORTOPTS LONGOPTS ARGS PROGNAME
+   unset generated_config updated
+   unset MACHINE DISTRO OEROOT TEMPLATECONF BBPATH TDROOT
+}
+
+# get command line options
+SHORTOPTS="h"
+LONGOPTS="help"
+
+ARGS=$(getopt --options $SHORTOPTS  \
+  --longoptions $LONGOPTS --name $PROGNAME -- "$@" )
```

Comment:
What do you think about adding options for specifying the machine name and distro, instead of requiring them to be specified as environment variables?

Reply:
Yeah, I can do that with all these options. I was thinking about adding this later anyhow to support setting local.conf options for things like the bootloader type, so these basic features might as well be in there too, especially if they're mostly going to have default options.

So right now, the only required option is going to be the machine name.

What do you think about using the machine name in the default build directory, like using `build-${MACHINE}` as the default?

Reply:
>  What do you think about using the machine name in the default build directory, like using build-${MACHINE} as the default?

I'd rather stick with just `build`.  We're supposed to be able to support building for different machines in the same workspace, and what I've seen with defaulting to separate workspaces per machine is that errors creep in that break that support.